### PR TITLE
Issuefixes

### LINKED ISSUE #487: check values in cfg files against .yaml files
currently, values of parameters defined in modules of cfg files are not checked. For example, if there are have many `choices` for a parameter in the yaml file, the code never checks that the value in the cfg file is part of the list present in the corresponding `defaults.yaml` file.

example from clustrmsd:
```
criterion:
  default: "maxclust"
  type: "string"
  ...
  choices:
  - distance
  - maxclust
```
Discussed with @joaomcteixeira.

Related to #333

### LINKED ISSUE #575: Booleans with capitals
**Describe the bug**

Booleans write with a capital letter cause an error when running haddock3.
It seems to come from the toml support that requires only small letters.
This is however not consistent with what we have in the default yaml files.

**To Reproduce**
Steps to reproduce the behavior, if relevant, provide the input that triggered the error:

Define a boolean starting with a capital letter, e.g.

randomremoval = False

**Expected behavior**

It is a rather trivial thing - I would think that raising an error for this is not necessary and the booleans can be transformed internally to lower case.

### LINKED ISSUE #578: Adding the config file to the data dir of a run
**Is your feature request related to a problem? Please describe.**

It would be nice and useful to copy the original config file of a run to the data dir of the run.
Currently we don't store it in the run which is not optimal for reproducibility.

**Describe the solution you'd like**

Simply copy the original cfg file to the run/data directory

### PR DESCRIPTION
**Pull request by VGPReys**
You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and that you comply with the following criteria:

- [X] You have sticked to Python. Please talk to us before adding other programming languages to HADDOCK3
- [ ] Your PR is about CNS
- [X] Your code is well documented: proper docstrings and explanatory comments for those tricky parts
- [X] You structured the code into small functions as much as possible. You can use classes if there is a (state) purpose
- [X] Your code follows our coding style
- [x] You wrote tests for the new code
- [x] `tox` tests pass. *Run `tox` command inside the repository folder*
- [x] `-test.cfg` examples execute without errors. *Inside `examples/` run `python run_tests.py -b`*
- [X] PR does not add any dependencies, unless permission granted by the HADDOCK team
- [X] PR does not break licensing
- [ ] Your PR is about writing documentation for already existing code :fire:
- [x] Your PR is about writing tests for already existing code :godmode:

---
This PR aims at closing three issues:
- With the addition of a regular expression that checks the presence of upper case True and False in the .toml input file and lowering them on-the-fly, this closes #575 
- Also by checking values types and ranges/choices, by comparing values present in the defaults.yaml and parameter =  value present in the user provided .cfg file, this closes #487 
- Finally, a bunch of copies of the workflow haddock3 input files are now written in the `data` directory of the run. This includes the `original inputed file`, the `regular expression cleaned file` and the `enhanced full parameter file`, which includes all default parameters in addition to the ones provided by the user. This closes #578 

Accordingly, additionnal tests were added, while some other test were modified to fit the new implementation.

Finally, `run_tests.py` ran without errors and the comparisons by `compare-runs.py` returned only `No errors found - OKAY!`.

### CODE REVIEW

File: src/haddock/gear/prepare_run.py [hunk 1]
```diff
@@ -75,6 +78,22 @@
...
+    'float': (float, int,),
+    'double': (float,),
+    'string': (str,),
+    'str': (str,),
+    'list': (list,),
+    'array': (list,),
+    'path': (str, Path, PosixPath, EmptyPath,),
+    'file': (str, Path, PosixPath, EmptyPath,),
+    }
+
```
Comment (mgiulini) [inline src/haddock/gear/prepare_run.py:96]: could you add a comment explaining this?

File: src/haddock/gear/prepare_run.py [hunk 2]
```diff
@@ -96,7 +115,7 @@ def wrapper(*args, **kwargs):
 
 
 @lru_cache
-def _read_defaults(module_name):
+def _read_defaults(module_name, default_only=True):
```
Comment (mgiulini) [inline src/haddock/gear/prepare_run.py:118]: what is default_only? could you write its meaning in the docstring?

File: src/haddock/gear/prepare_run.py [hunk 3]
```diff
@@ -315,6 +340,74 @@ def setup_run(
...
+
+    # Add README to help user
+    with open(Path(confpaths, 'README.txt'), 'w') as f:
+        f.write(f"{'#'*80}\n# Information about configuration "
+                f"files present in the same directory #\n{'#'*80}\n")
+        for confname in list_save_conf:
+            if confname not in added_files.keys():
+                continue
+            f.write(f'"{added_files[confname]}": {infofile[confname]}\n')
+
```
Comment (mgiulini) [inline src/haddock/gear/prepare_run.py:409]: I would return something here, like the paths to the saved file

File: src/haddock/gear/prepare_run.py [hunk 4]
```diff
@@ -404,6 +497,144 @@ def pretty_print(match):
                 )
             raise ConfigurationError(_msg)
 
+        # Now that existance of the parameter was checked,
```
Comment (mgiulini) [inline src/haddock/gear/prepare_run.py:500]:
```suggestion
        # Now that existence of the parameter was checked,
```

File: src/haddock/gear/yaml2cfg.py [hunk 5]
```diff
@@ -119,13 +119,32 @@ def _yaml2cfg_text(ycfg, module, explevel):
...
+        Set the return value of this function; if True (default value), only
+        returns default values, else return the fully loaded configuration file
+
+    Return
+    ------
+    ycfg : dict
+        The full default configuration file as a dict
+    OR
+    cfg : dict
+        A dictionnary containing only the default parameters values
```
Comment (mgiulini) [inline src/haddock/gear/yaml2cfg.py:139]:
```suggestion
        A dictionary containing only the default parameters values
```

Comment (mgiulini) [inline src/haddock/gear/yaml2cfg.py:139]: is there any particular reason to have two different dict objects in output?

**Review by mgiulini — COMMENTED** [review]
added a few comments, I will test it asap!

File: src/haddock/gear/yaml2cfg.py (hunk 5 above)
Comment (VGPReys) [reply src/haddock/gear/yaml2cfg.py:139]: There is the original one `cfg` that only contains the ['default'] value and used in other parts of the code.
While the `ycfg` one contrains verything present in the default.yaml file loaded as  a dict, used for the analysis of the agreement (type, range, choices, sizes) between user provided configuration and authorized one.

File: src/haddock/gear/prepare_run.py (hunk 2 above)
Comment (VGPReys) [reply src/haddock/gear/prepare_run.py:118]: Good idea !

**Comment by rvhonorato** [comment]
Sorry I have absolutely no idea about this part of the code, will trust your judgment 🤞🏽

**Review by mgiulini — APPROVED** [review]
Runs nicely on a couple of examples, great addition @VGPReys!

**Pull request merged**